subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/processors/RunSortedBySizeTestClassProcessor.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.testing.processors;
+
+import org.gradle.api.internal.tasks.testing.TestClassProcessor;
+import org.gradle.api.internal.tasks.testing.TestClassRunInfo;
+import org.gradle.api.internal.tasks.testing.TestResultProcessor;
+
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.LinkedHashSet;
+import java.util.LinkedList;
+import java.util.Map;
+
+/**
+ * This processor uses test class size information to resort test classes by
+ * size in descending order before passing them on for processing.
+ *
+ * Tests processed without size information are at the end behind known classes.
+ * If no sizes at all are known beforehand, this will not change the test order.
+ */
+public class RunSortedBySizeTestClassProcessor implements TestClassProcessor {
+	private final TestClassProcessor delegate;
+	private final Map<String, Long> sizeByTestClassName;
+
+	private final LinkedList<TestClassRunInfo> knownTestClasses = new LinkedList<TestClassRunInfo>();
```

Comment:
Please never use `LinkedList`. In 99.42% of the cases you do not need it.

Reply:
First: Thanks for the quick and extensive feedback!

I have changed the type (this was a leftover from when I wanted to add the unknown test cases in the middle of the known ones - see also the issue with the comment).